nep-12.mediawiki

```diff
@@ -0,0 +1,2130 @@
+    case NO_PROVIDER:
+      console.log('No provider available.');
+      break;
+    case CONNECTION_DENIED:
+      console.log('The user rejected the request to connect with your dApp');
+      break;
+  }
+});
+</pre>
+
+
+'''Provider Request Handling'''
+
+Upon receiving this request, return:
+- The address and label of account currently selected by the user in the wallet interface
+
+Example
+<pre>
+{
+  address: 'AeysVbKWiLSuSDhg7DTzUdDyYYKfgjojru',
+  label: 'My Spending Wallet'
+}
+</pre>
+
+
+====getPublicKey====
+Return the public key of the Account that is currently connected to the dApp.
+
+
+'''Method Interface'''
+
+<pre>
+function getPublicKey(): Promise<{ address: string, publicKey: string }>
+</pre>
+
+
+'''Input arguments'''
+
+None
+
+
+'''Success return value'''
+
+<pre>
+interface Account {
+  address: string; // Address of the connected account
+  publicKey: string; // Public key of the connected account
+}
+</pre>
+
+
+'''Error return value'''
+
+<pre>
+interface Error {
+  type: string; // `NO_PROVIDER`|`CONNECTION_DENIED`
+  description: string;
+  data: string;
+}
+</pre>
+
+
+'''Example'''
+
+<pre>
+getPublicKey()
+.then((publicKeyData: PublicKeyData) => {
+  const {
+    address,
+    publicKey,
+  } = publicKeyData;
+
+  console.log('Account address: ' + address);
+  console.log('Account public key: ' + publicKey);
+})
+.catch(({type: string, description: string, data: any}) => {
+  switch(type) {
+    case NO_PROVIDER:
+      console.log('No provider available.');
+      break;
+    case CONNECTION_DENIED:
+      console.log('The user rejected the request to connect with your dApp');
+      break;
+  }
+});
+</pre>
+
+
+'''Provider Request Handling'''
+
+Upon receiving this request, return:
+- The address and public key of account currently selected by the user in the wallet interface
+
+Example
+<pre>
+{
+  address: 'ATUaTd3LA4kZiyB6it9fdb5oJpZYMBF4DX',
+  publicKey: '03fa41b6ff75ebeff8464556629cfceae7402f5d815626a7a6542f786974b942e0'
+}
+</pre>
+
+
+====getBalance====
+Return balance of a specific asset for the given account.
+
+If the asset is omitted from a request to MainNet, all asset and token balances will be returned.
+
+
+'''Method Interface'''
+
+<pre>
+function getBalance(args: GetBalanceArgs): Promise<BalanceResults>
+</pre>
+
+
+'''Input arguments'''
+
+<pre>
+interface GetBalanceArgs {
+  params: BalanceRequest|BalanceRequest[];
+  network?: string - Network to submit this request to. If omitted, will default to network the wallet is currently set to.
+}
+
+interface BalanceRequest {
+  address: string; // Address to check balance(s)
+  assets?: string[]; // Asset IDs or script hashes to check balance.
+  fetchUTXO?: boolean; // Fetches to UTXO data for NEO and/or GAS if attribute is 'true'
+}
+</pre>
+
+
+'''Success return value'''
+
+<pre>
+interface BalanceResults {
+  [address: string]: Balance[];
+}
+
+interface Balance {
+  assetID: string;
+  symbol: string;
+  amount: string;
```

Comment:
`amount: string` cannot be correct (or at least, not precise)
Anyway, I suggest the following here. Return `amount` in integer only, and add another field for decimal digits of asset. Usually, digits will be 8 (besides NEO asset, which is 0).
By adopting floating point here, like I saw some examples below, it may incentive practices that are very dangerous, because dealing with such small fractions in non-precision standard, will certainly lead to errors in some implementations. integer is guaranteed on every existing language.

Reply:
A common problem for a lot of developers coming into the ecosystem was the confusion around using this integer + decimal combination to express values. As you mentioned, unless it's well understood how to handle this conversion properly, errors may occur in implementation. This is why we are looking to offload this process to wallet providers, so that the conversion can be handled in fewer places. Thus making the total number of implementations with possible errors much smaller.

Reply:
Ok,if string is the final decision, this should be informed clearly on text, because seeing "amount : string" is very strange to me... if it was explained, I would agree. Its understandable.